src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts

```diff
@@ -373,12 +373,20 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 			this._tasksReconnected = true;
 			return;
 		}
-		this._getTaskSystem();
 		this.getWorkspaceTasks().then(async () => {
 			this._tasksReconnected = await this._reconnectTasks();
 		});
 	}
 
+	private async _waitForTaskSystem(): Promise<void> {
+		if (this.hasTaskSystemInfo) {
+			return;
+		}
+		// Wait until we have task system info (the extension host and workspace folders are available).
+		this._logService.trace('RunBuildTask: Awaiting task system info.');
+		await Event.toPromise(Event.once(this.onDidChangeTaskSystemInfo));
```

Comment:
This could introduce a bug in the future if there is any async code in `registerTaskSystem`. How about instead you set a property on startup which does the same thing?

```ts
readonly _whenTaskSystemReady = Event.toPromise(this.onDidChangeTaskSystemInfo);
```

Then just await that instead?